fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/test/FDBDatabaseExtension.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * FDBDatabaseFactoryExtension.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.test;
+
+import com.apple.foundationdb.FDB;
+import com.apple.foundationdb.record.provider.foundationdb.APIVersion;
+import com.apple.foundationdb.record.provider.foundationdb.BlockingInAsyncDetection;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabase;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseFactory;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseFactoryImpl;
+import com.apple.foundationdb.test.TestExecutors;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+import java.util.concurrent.Executor;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Extension that allows the user to specify the database. It ensures that FDB has been properly initialized
+ * and that the {@link FDBDatabase} object used during a test does not leak between test runs. This registers
+ * call backs that run before and after tests, so it is suggested that users use the {@link org.junit.jupiter.api.extension.RegisterExtension}
+ * annotation to ensure that those callbacks run. Like so:
+ *
+ * <pre>{@code
+ *     @RegisterExtension
+ *     static final FDBDatabaseExtension = new FDBDatabaseExtension();
+ * }</pre>
+ */
+public class FDBDatabaseExtension implements BeforeEachCallback, AfterEachCallback {
```

Comment:
The one in `fdb-extensions` doesn't rely on any Record Layer abstractions. This one, in addition to initializing FDB, also creates a new factory and database object for each test, which allows each test to modify what would normally be global state. In theory, we could re-use `TestDatabaseExtension` here to initialize FDB, but then we would never exercise the `FDBDatabaseFactory`'s `initFDB` method in tests, which seems like it would be a bit of a miss.